### PR TITLE
Inital commit of job_manager

### PR DESCRIPTION
**Pull request by papagalu**
More to come

### CODE REVIEW

File: scripts/job_manager.ps1 [hunk 1]
```diff
@@ -0,0 +1,94 @@
+class MyJobManager {
```
Comment (ader1990) [inline scripts/job_manager.ps1]: Change the name to a more suitable name like: PSJobManager

File: scripts/job_manager.ps1 [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    [String] $Name = "JobManager"
```
Comment (ader1990) [inline scripts/job_manager.ps1]: PSJobManager

File: scripts/job_manager.ps1 [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    [Hashtable] $Jobs = @{}
+
+    JobManager () {
+    }
+
+    [void] AddJob ($Tracker, $ScriptBlock, $ArgumentList, $InitScript) {
+        $job = Start-Job -ScriptBlock $ScriptBlock `
+                         -ArgumentList $ArgumentList `
+                         -InitializationScript $InitScript
+        Start-Sleep -Milliseconds 100
```
Comment (ader1990) [inline scripts/job_manager.ps1]: why do you need this sleep here?

File: scripts/job_manager.ps1 [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+        if ($job.State -ne "Running" ) {
```
Comment (ader1990) [inline scripts/job_manager.ps1]: are there other states you need to keep track of? what are the possible states of a job?

File: scripts/job_manager.ps1 [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+            $this.Jobs[$Tracker].AddLast($job)
+        } else {
+            $this.Jobs[$Tracker] = New-Object Collections.Generic.LinkedList[object]
+            $this.Jobs[$Tracker].AddLast($job)
+        }
+    }
+
+    [void] RemoveTracker ($Tracker) {
+        Write-Host "Removing jobs from tracker: $Tracker"
+        Start-Sleep 5
```
Comment (ader1990) [inline scripts/job_manager.ps1]: ditto

Comment (ader1990) [reply scripts/job_manager.ps1]: From a general perspective, start sleeps are an obvious lack of retries

File: scripts/job_manager.ps1 [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+            $current = $current.Next
+        }
+        $this.Jobs.Remove($Tracker)
+    }
+
+    [Collections.Generic.LinkedList[object]] GetJobsFromTracker ($Tracker) {
+        if ($this.Jobs.Contains($Tracker)) {
+            return $this.Jobs[$Tracker]
+        } else {
+            Write-Host "no tracker"
```
Comment (ader1990) [inline scripts/job_manager.ps1]: Add a more verbose message like: "Tracker $Tracker not found"

File: scripts/job_manager.ps1 [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+            return $null
+        }
+    }
+
+    [String] WaitForJobCompletion ($Tracker, $Timeout ) {
+        Write-Host "Waiting for Jobs to finish in Tracker : $Tracker"
```
Comment (ader1990) [inline scripts/job_manager.ps1]: Tracker:

File: scripts/job_manager.ps1 (hunk 6 above)
Comment (ader1990) [reply scripts/job_manager.ps1]: Add a synopsis for each method what a tracker and more, if you add new concepts

File: scripts/job_manager.ps1 [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+
+    [Array] GetJobOutputs ($Tracker) {
+        Write-Host "Retriveing Output from Jobs in Tracker : $Tracker"
+        $asd = ""
+        [Console]::InputEncoding = New-Object Text.UTF8Encoding $false 
+        $results = @()
+        $current = $this.Jobs[$Tracker].First
+        while(-not ($current -eq $null)) {
+            $asd = $current.Value | Receive-Job -Keep
+            $results += $asd
```
Comment (ader1990) [inline scripts/job_manager.ps1]: what is asd?

File: scripts/job_manager.ps1 [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+                    $status = "not running"
+                }
+                $current = $current.Next
+            }
+        }
+       return $status
+    }
+
+    [void] StopJobsFromTracker ($Tracker) {
+        Write-Host "Stoping Jobs From Tracker : $Tracker"
```
Comment (ader1990) [inline scripts/job_manager.ps1]: ditto

File: scripts/job_manager.ps1 [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+            $Timeout--
+            $current = $this.Jobs[$Tracker].First
+            Write-Host "time is $timeout"
+            while(-not ($current -eq $null)) {
+                if ($current.Value.State -eq "Running") {
+                    $status = "running"
+                    Start-Sleep 1
+                    break
+                } else {
+                    $status = "not running"
```
Comment (ader1990) [inline scripts/job_manager.ps1]: These statuses need to be added in an enumeration

File: scripts/job_manager.ps1 [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+            return $this.Jobs[$Tracker]
+        } else {
+            Write-Host "no tracker"
+            return $null
+        }
+    }
+
+    [String] WaitForJobCompletion ($Tracker, $Timeout ) {
+        Write-Host "Waiting for Jobs to finish in Tracker : $Tracker"
+        $status = "running"
```
Comment (ader1990) [inline scripts/job_manager.ps1]: look bellow

**Review by ader1990 — CHANGES_REQUESTED** [review]

File: scripts/job_manager.ps1 (hunk 3 above)
Comment (papagalu) [reply scripts/job_manager.ps1]: when the job starts, it is not in the Running state immediately.

File: scripts/job_manager.ps1 (hunk 5 above)
Comment (papagalu) [reply scripts/job_manager.ps1]: previously i kill the lisa process, and the job doesn't stop instantly
will try to stop the job first

File: scripts/job_manager.ps1 (hunk 11 above)
Comment (papagalu) [reply scripts/job_manager.ps1]: when waiting for job, it only makes sense to wait for them if they are in the running state
no need to change this method

File: scripts/job_manager.ps1 (hunk 4 above)
Comment (papagalu) [reply scripts/job_manager.ps1]: the job should be in the running state; if something bad happens, we would like to see the output and not add the job to the tracker regardless of the state the job is in.

File: scripts/job_manager.ps1 [hunk 12]
```diff
@@ -0,0 +1,143 @@
+<#
+.SYNOPSIS
+
+class to ease job management
```
Comment (ader1990) [inline scripts/job_manager.ps1]: Class to ease PowerShell Job management.

File: scripts/job_manager.ps1 [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+Adds a job to a "Tracker"; a Tracker is a set of similar jobs that should be
```
Comment (ader1990) [inline scripts/job_manager.ps1]: You can add a PowerShell Job to a Tracker. A Tracker is a label for a set of jobs that you want to manage in the same way (start, stop, wait).

File: scripts/job_manager.ps1 [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+operated in parallel(start, stop, wait, etc)
+
+you can easily manipulate a single job by adding it to a unique tracker.
```
Comment (ader1990) [inline scripts/job_manager.ps1]: You*

File: scripts/job_manager.ps1 [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+    <#
+    .SYNOPSIS
+
+    Creates a Job and adds it to a Tracker
+    #>
+    [void] AddJob ($Tracker, $ScriptBlock, $ArgumentList, $InitScript) {
+        $job = Start-Job -ScriptBlock $ScriptBlock `
+                         -ArgumentList $ArgumentList `
+                         -InitializationScript $InitScript
+        $Timeout = 10
```
Comment (ader1990) [inline scripts/job_manager.ps1]: $timeout

File: scripts/job_manager.ps1 [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+        while ($Timeout -gt 0) {
+            if ($job.State -eq "Running") {
+                break;
+            } else {
+                Start-Sleep 1
+                $Timeout -= 1
+            }
+        }
+        if ($job.State -ne "Running" ) {
```
Comment (ader1990) [inline scripts/job_manager.ps1]: What if it has finished already?

File: scripts/job_manager.ps1 [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+    Removes all jobs in the tracker and finally removes the tracker itself
+    #>
+    [void] RemoveTracker ($Tracker) {
+        Write-Host "Removing jobs from tracker: $Tracker"
+        $current = $this.Jobs[$Tracker].First
+        while (-not ($current -eq $null)) {
+            $Timeout = 5
+            while ($Timeout -gt 0) {
+                try {
+                    Remove-Job $current.Value
```
Comment (ader1990) [inline scripts/job_manager.ps1]: Stop-Job should also be done before.

File: scripts/job_manager.ps1 [hunk 18]
```diff
@@ -0,0 +1,143 @@
...
+    <#
+    .SYNOPSIS
+
+    Removes all jobs in the tracker and finally removes the tracker itself
+    #>
+    [void] RemoveTracker ($Tracker) {
+        Write-Host "Removing jobs from tracker: $Tracker"
+        $current = $this.Jobs[$Tracker].First
+        while (-not ($current -eq $null)) {
+            $Timeout = 5
```
Comment (ader1990) [inline scripts/job_manager.ps1]: $timeout

File: scripts/job_manager.ps1 [hunk 19]
```diff
@@ -0,0 +1,143 @@
...
+                $current = $current.Next
+            }
+        }
+       return $status
+    }
+    
+    <#
+    .SYSNOPSIS
+
+    Forcefully stop the LISA process(if something goes wrong, it hangs)
```
Comment (ader1990) [inline scripts/job_manager.ps1]: this is a generic job manager, remove LISA from here

File: scripts/job_manager.ps1 [hunk 20]
```diff
@@ -0,0 +1,143 @@
...
+            Write-Host "Stop job with pid : $mypid"
+            Stop-Process $myPid
+            $current = $current.Next
+        }
+    }
+
+    <#
+    .SYSNOPSIS
+
+    Returns an array with the jobresults in the specified Tracker
```
Comment (ader1990) [inline scripts/job_manager.ps1]: job results for the specified Topic.

**Review by ader1990 — CHANGES_REQUESTED** [review]

File: scripts/job_manager.ps1 [hunk 21]
```diff
@@ -0,0 +1,155 @@
...
+    [Array] $RunningStates = @("AtBreakpoint", "Blocked", "Disconnected", "Running", "Stopping", "Suspended", "Suspending")
+    [Array] $FinishedStates = @("Completed", "Failed", "Stopped")
+
+    JobManager () {
+    }
+
+    <#
+    .SYNOPSIS
+
+    Creates a Job and adds it to a Topic
```
Comment (ader1990) [inline scripts/job_manager.ps1]: Starts a job and adds it to a Topic.
see the . at the end

File: scripts/job_manager.ps1 [hunk 22]
```diff
@@ -0,0 +1,155 @@
...
+        }
+    }
+
+    <#
+    .SYSNOPSIS
+
+    Job results for the specified Topic
+    #>
+    [Array] GetJobOutputs ($Topic) {
+        Write-Host "Retriveing Output from Jobs in Topic: $Topic"
```
Comment (ader1990) [inline scripts/job_manager.ps1]: Retrieving*

File: scripts/job_manager.ps1 [hunk 23]
```diff
@@ -0,0 +1,155 @@
...
+    
+    Waits for all jobs in a tracker to be in non-Running state.
+    #>
+    [String] WaitForJobCompletion ($Topic, $Timeout ) {
+        Write-Host "Waiting for Jobs to finish in Topic: $Topic"
+        $status = "running"
+        while (($status -eq "running") -and ($Timeout -gt 0)) {
+            $Timeout -= 1
+            $current = $this.Jobs[$Topic].First
+            Write-Host "time is $timeout"
```
Comment (ader1990) [inline scripts/job_manager.ps1]: Waiting for job to stop <JobName>...

File: scripts/job_manager.ps1 [hunk 24]
```diff
@@ -0,0 +1,155 @@
...
+            while(-not ($current -eq $null)) {
+                if ($current.Value.State -eq "Running") {
```
Comment (ader1990) [inline scripts/job_manager.ps1]: Use the jobstate enumeration instead of the string Running.

File: scripts/job_manager.ps1 [hunk 25]
```diff
@@ -0,0 +1,155 @@
...
+                    $status = "running"
+                    Start-Sleep 1
+                    break
+                } else {
+                    $status = "not running"
+                }
+                $current = $current.Next
+            }
+        }
+       return $status
```
Comment (ader1990) [inline scripts/job_manager.ps1:124]: return the actual status here.

**Review by ader1990 — CHANGES_REQUESTED** [review]

**Pull request merged**